### PR TITLE
Boosted

### PR DESCRIPTION
**Pull request by mmerce**

### CODE REVIEW

File: bigml/boostedtree.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            0 - LAST_PREDICTION: the last issued prediction is returned.
+            1 - PROPORTIONAL: we consider all possible outcomes and create
+                              an average prediction.
+        """
+
+        if path is None:
+            path = []
+        if missing_strategy == PROPORTIONAL:
+            return self.predict_proportional(input_data, path=path)
+
```
Comment (jaor) [inline bigml/boostedtree.py]: silly, but i guess pylint will complain about this blank line

File: bigml/model.py [hunk 2]
```diff
@@ -333,6 +361,18 @@ def predict(self, input_data, by_name=True,
         prediction = self.tree.predict(input_data,
                                        missing_strategy=missing_strategy)
 
+        if self.boosting is not None and missing_strategy == PROPORTIONAL:
+            # output has to be recomputed and comes in a different format
+            g_sum, h_sum, population, path = prediction
+            prediction = Prediction(
+                - g_sum / (h_sum +  self.boosting.get("lambda", 1)),
```
Comment (jaor) [inline bigml/model.py]: so h_sum can never be -1 here, right?

File: bigml/multivote.py [hunk 3]
```diff
@@ -384,10 +422,42 @@ def combine(self, method=DEFAULT_METHOD, with_confidence=False,
...
+                prediction, prediction_info = sorted( \
+                    predictions.items(), key=lambda(x): \
+                    (- x[1]["probability"], x[1]["order"]))[0]
+                confidence = prediction_info["probability"]
+                if with_confidence:
+                    return prediction, confidence
+                elif add_confidence:
+                    return {"prediction": prediction, "confidence": confidence}
+                else:
+                    return prediction
```
Comment (jaor) [inline bigml/multivote.py]: for symmetry with the other branch of the `if`, and to make this big method shorter, a possibility is to put these ~20 lines in a little function... but maybe you have reasons not to...

File: bigml/tests/compare_predictions_steps.py [hunk 4]
```diff
@@ -220,20 +227,33 @@ def the_local_prediction_is(step, prediction):
...
     except AttributeError:
-        local_model = world.local_ensemble.multi_model.models[0]
-        if local_model.tree.regression:
-            local_prediction = round(float(local_prediction), 4)
-            prediction = round(float(prediction), 4)
+        local_model = world.local_ensemble
+        if local_model.regression:
+            assert_almost_equal(local_prediction, float(prediction), places=5)
+    if local_prediction != prediction:
+        assert False, "found: %s, expected %s" % (local_prediction, prediction)
```
Comment (jaor) [inline bigml/tests/compare_predictions_steps.py]: `assert_equal(local_prediction, prediction)`

File: bigml/model.py (hunk 2 above)
Comment (mmerce) [inline bigml/model.py]: apparently not

File: bigml/multivote.py (hunk 3 above)
Comment (mmerce) [inline bigml/multivote.py]: no special reason. I'll have a look.

File: bigml/tests/compare_predictions_steps.py [hunk 5]
```diff
@@ -220,20 +227,33 @@ def the_local_prediction_is(step, prediction):
...
+        local_prediction = world.local_prediction[0]
+    elif isinstance(world.local_prediction, dict):
+        local_prediction = world.local_prediction['prediction']
     else:
+        local_prediction = world.local_prediction
+    local_model = world.local_ensemble
+    if local_model.regression:
+        assert_almost_equal(local_prediction, float(prediction), places=5)
+    elif local_prediction != prediction:
         assert False, "found: %s, expected %s" % (local_prediction, prediction)
```
Comment (jaor) [inline bigml/tests/compare_predictions_steps.py]: while we are around, here we could also use `assert_equal`

Comment (mmerce) [inline bigml/tests/compare_predictions_steps.py]: np

File: bigml/tests/compare_predictions_steps.py (hunk 4 above)
Comment (mmerce) [inline bigml/tests/compare_predictions_steps.py]: ok

File: bigml/tests/create_ensemble_steps.py [hunk 6]
```diff
@@ -74,15 +76,17 @@ def the_ensemble_is_finished_in_less_than(step, secs):
 #@step(r'I create a local Ensemble$')
 def create_local_ensemble(step):
     world.local_ensemble = Ensemble(world.ensemble_id, world.api)
-    world.local_model = Model(world.local_ensemble.model_ids[0])
+    world.local_model = Model(world.local_ensemble.model_ids[0], world.api)
 
 #@step(r'I create a local Ensemble with the last (\d+) models$')
 def create_local_ensemble_with_list(step, number_of_models):
-    world.local_ensemble = Ensemble(world.models[-int(number_of_models):], world.api)
+    world.local_ensemble = Ensemble(world.models[-int(number_of_models):],
```
Comment (jaor) [inline bigml/tests/create_ensemble_steps.py]: github is displaying this line as if it had lots of whitespace at the end... probably it's just github...

File: bigml/tests/create_ensemble_steps.py [hunk 7]
```diff
@@ -74,15 +76,17 @@ def the_ensemble_is_finished_in_less_than(step, secs):
...
+                                    world.api)
 
 #@step(r'I create a local Ensemble with the last (\d+) local models$')
 def create_local_ensemble_with_list_of_local_models(step, number_of_models):
-    local_models = [Model(model) for model in world.models[-int(number_of_models):]]
+    local_models = [Model(model) for model in
```
Comment (jaor) [inline bigml/tests/create_ensemble_steps.py]: github is displaying this line as if it had lots of whitespace at the end... probably it's just github...

Comment (mmerce) [inline bigml/tests/create_ensemble_steps.py]: Strange, my editor removes them. I'll check.

File: bigml/tests/create_prediction_steps.py [hunk 8]
```diff
@@ -107,6 +107,20 @@ def i_create_an_ensemble_prediction(step, data=None):
...
+def i_create_an_ensemble_proportional_prediction(step, data=None):
+    if data is None:
+        data = "{}"
+    ensemble = world.ensemble['resource']
+    data = json.loads(data)
+    resource = world.api.create_prediction(ensemble,
+                                           data,
+                                           {"missing_strategy": 1})
+    world.status = resource['code']
+    assert world.status == HTTP_CREATED
```
Comment (jaor) [inline bigml/tests/create_prediction_steps.py]: `assert_equal`

**Comment by mmerce** [comment]
closing this PR to merge the boosted2 branch that contains this code and open another comparing the boosted branch to this.

**Pull request closed without merging**